app/assets/src/components/visualizations/heatmap/Heatmap.js

```diff
@@ -400,6 +423,140 @@ export default class Heatmap {
+          this.spacePressed = false;
+        }
+      });
+
+    this.svg
+      .on("mousedown", () => {
+        this.mouseDown = true;
+        if (this.spacePressed) {
+          this.mouseX = d3.event.clientX;
+          this.mouseY = d3.event.clientY;
+          d3.event.preventDefault();
+        }
+      })
+      .on("mousemove", this.drag.bind(this))
+      .on("mouseup", () => {
+        this.mouseDown = false;
+      });
+
+    // Set up scrolling behavior.
+    this.g.on("wheel.zoom", this.scroll.bind(this));
+  }
+
+  drag() {
+    if (this.mouseDown && this.spacePressed) {
+      this.pan(d3.event.clientX - this.mouseX, d3.event.clientY - this.mouseY);
+      this.mouseX = d3.event.clientX;
+      this.mouseY = d3.event.clientY;
+    }
+  }
+
+  scroll() {
+    this.pan(d3.event.wheelDeltaX, d3.event.wheelDeltaY);
+    d3.event.stopPropagation();
+  }
+
+  pan(deltaX, deltaY) {
+    // Define the scrolling boundaries for the svg.
+    // Upper limits are determined by the difference between the container and svg sizes,
+    // scaled by the zoom factor.
+    let containerWidth = this.container[0][0].offsetWidth,
+      containerHeight = this.container[0][0].offsetHeight;
+    let xScrollMax =
+        (containerWidth - this.svg.attr("width")) / this.options.zoom,
+      yScrollMax =
+        (containerHeight - this.svg.attr("height")) / this.options.zoom;
+
+    // Translating the svg:
+    let gCurrentTranslate = d3.transform(this.g.attr("transform")).translate;
+    // Limit translation by the boundaries set for the svg.
+    let dx = Math.min(
+      this.options.marginLeft,
+      Math.max(deltaX + gCurrentTranslate[0], xScrollMax)
+    );
+    let dy = Math.min(
+      this.options.marginTop,
+      Math.max(deltaY + gCurrentTranslate[1], yScrollMax)
+    );
+    this.g.attr("transform", `translate(${[dx, dy]})`);
+
+    // Translating the row labels in the opposite x direction of the svg.
+    let rowLabelsCurrent = d3.transform(this.gRowLabels.attr("transform"))
+      .translate;
+    let labelsDx = clamp(
+      0,
+      -xScrollMax + this.options.marginLeft,
+      rowLabelsCurrent[0] - deltaX
+    );
+    this.placeRowLabels(labelsDx);
+
+    // Translating the metadata labels in the opposite x direction of the svg (same as row labels).
+    // Don't include the transition animation while rendering.
+    this.renderColumnMetadataLabels(labelsDx, false);
+    // Translate the "Add Metadata" link (same as row labels).
+    this.renderColumnMetadataAddLink(labelsDx);
+
+    // Translating the column and metadata labels in the opposite y direction of the svg.
+    let columnLabelsCurrent = d3.transform(this.gColumnLabels.attr("transform"))
+      .translate;
+    let labelsDy = clamp(
+      this.columnLabelsHeight,
+      this.columnLabelsHeight + this.options.marginTop - yScrollMax,
+      columnLabelsCurrent[1] - deltaY
+    );
+    this.placeColumnAndMetadataLabels(labelsDy);
```

Comment:
Can we rename this to `placeColumnLabelAndMetadataContainers`? Otherwise, it's confusing why we're calling `renderColumnMetadataLabels(labelsDx)` above and `placeColumnAndMetadataLabels(labelsDy)` here. You would expect both dx and dy to be passed into the "place" method and nothing to be passed to the "render" method based on the naming. The reason is that the dy translation is applied to the container, while dx is applied to each individual metadata label, and the renaming makes that a little more clear. (also, `renderColumnMetadataLabels` both renders and places)

Also, rename `placeRowLabels` to `placeRowLabelContainers` for consistency.

Finally, I think a further improvement would be to split up the renderColumnMetadataLabel and placeColumnMetadataLabel logic into separate function and only call the place function here. So that you only execute the exact logic that you need. However, we can defer that to a later time, to avoid endlessly refactoring code.